signer/signer.go

```diff
@@ -0,0 +1,134 @@
+package signer
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+
+	msgsigner "github.com/form3tech-oss/go-http-message-signatures"
+	"github.com/form3tech-oss/http-message-signing-proxy/config"
+	"github.com/form3tech-oss/http-message-signing-proxy/proxy"
+)
+
+const (
+	requestTargetHeaderKey = "(request-target)"
+	digestHeaderKey        = "digest"
+)
+
+type DataError = msgsigner.DataError
+
+type requestSigner struct {
+	*msgsigner.MessageSigner
+	headerConfig config.HeadersConfig
+}
+
+func NewRequestSigner(cfg config.SignerConfig) (proxy.RequestSigner, error) {
+	key, err := loadKey(cfg.KeyFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	signatureHashAlgo, err := getHashAlgo(cfg.SignatureHashAlgo)
+	if err != nil {
+		return nil, err
+	}
+
+	digestHashAlgo, err := getHashAlgo(cfg.SignatureHashAlgo)
+	if err != nil {
+		return nil, err
+	}
+
+	signer, err := msgsigner.NewRSASigner(key, signatureHashAlgo)
+	if err != nil {
+		return nil, err
+	}
+
+	msgSigner, err := msgsigner.NewMessageSigner(digestHashAlgo, signer, cfg.KeyId, msgsigner.Signature)
+	if err != nil {
+		return nil, err
+	}
+
+	return &requestSigner{
+		MessageSigner: msgSigner,
+		headerConfig:  cfg.Headers,
+	}, err
+}
+
+func (rs *requestSigner) SignRequest(req *http.Request) (*http.Request, error) {
```

Comment:
just a question: what will happen if the request already contains `Signature` header?

Reply:
We will overwrite it

Reply:
Thank you. I was just curious. I'm not sure how far we should go with preventing users from doing silly things. I can imagine someone adds `signature` to required headers used for signing which would make verification fail.